### PR TITLE
datatype: reorganize layout calculation code

### PR DESCRIPTION
**Pull request by vtjnash**
The code felt a bit disorganized to me, especially after #32448, and with the upcoming desire to eventually merge #18632. I think this will make it clearer when adopting further changes by changing where some of the early-return occurs and thus also removing some indentation (hence also, diff readability may be improved by ignoring whitespace).

### CODE REVIEW

**Comment by vtjnash** [comment]
Seems unlikely to be noticeable, but just to check: @nanosoldier `runbenchmarks(ALL, vs=":master")

**Comment by vtjnash** [comment]
Seems unlikely to be noticeable, but just to check: @nanosoldier `runbenchmarks(ALL, vs=":master")`

File: src/datatype.c [hunk 1]
```diff
@@ -296,165 +296,183 @@ static int references_name(jl_value_t *p, jl_typename_t *name) JL_NOTSAFEPOINT
     return 0;
 }
 
+static void throw_ovf(int should_malloc, void *desc, jl_datatype_t* st, int offset)
+{
+    if (should_malloc)
```
Comment (JeffBezanson) [inline src/datatype.c]: `should_free`?

Comment (vtjnash) [reply src/datatype.c]: I suppose it could. It's currently just carrying the same name along.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/33724#issuecomment-554415070) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ebae3c7d902d4f79e9753c9cab14feb8f875ef5b/89b1698_vs_4d7701e/report.md). cc @ararslan

File: src/datatype.c (hunk 1 above)
Comment (vtjnash) [reply src/datatype.c]: Do you care?

Comment (JeffBezanson) [reply src/datatype.c]: Not so important I guess.

**Pull request merged**